### ISSUE TITLE
[Enhancement] Add ability to change host configuration for API hosts.

### ISSUE THREAD

**Issue by barefootdeveloper** [mockoon/mockoon#469]
**Is your feature request related to a problem? Please describe.**
Not a problem, but being able to restrict API availability to localhost would be nice.

**Describe the solution you'd like**
It would be useful to allow the underlying host configuration for the mock APIs to be configured. This would allow a user to restrict what interfaces the mock server would respond to.

For the interface I was thinking of a drop-down picking between All Hosts (0.0.0.0) and local host only (127.0.0.1) but it might be useful to expand it later to allow for manual entry in scenarios where a machine has multiple network cards.

I believe this will require changes to:

- commons/src/models/environment.model.ts - Add Host Property
- commons-server/src/libs/server.ts - Make use of host property, default to 0.0.0.0 if not defined.
- A new component to be referenced from mockoon/src/renderer/app/app.component.html

I didn't dig into the CLI tool, it may need changes there to allow for configuration of the host.

I was actually thinking of picking up this work since I think it's a nice value-add, but not really necessary since you can simply configure firewall rules to restrict outside access.

**Comment by barefootdeveloper** [comment]
I have the work for this largely completed on my local. There could be some more work to do on the port error warning since now that the hostname is configurable you can run two environments on the same port if their hostnames are different. This creates the possibility for some interesting scenarios where you can set up a remote environment that just proxies to your local environment but would give you the ability to monitor the calls separately or provide different mock values.

I was curious if anyone had any opinions on a drop down vs input value?

The drop down has the advantage of being simple, but it's also limited in that you can't insert IP addresses of specific network cards.

The input value has the advantage of being very flexible, but allows you to hurt yourself by putting in something invalid. Without any validation effort the application to it's credit handles this pretty well.

Dropdown
![Snag_179218e1](https://user-images.githubusercontent.com/3580448/117449721-4b243880-af0e-11eb-900d-555f3497c23d.png)

Input
![Snag_17922814](https://user-images.githubusercontent.com/3580448/117449731-4d869280-af0e-11eb-862f-169e3b103ffb.png)

Error whenever illogical input is attempted to be started
![Snag_17923c57](https://user-images.githubusercontent.com/3580448/117449741-50818300-af0e-11eb-9bb1-d5c51ad6c468.png)

**Comment by 255kb** [comment]
That would be indeed an interesting addition. 
However, multiple things are coming to my mind: 

- I don't think this setting is shareable between coworkers, so, I feel like it would need to be excluded from the export option. Which is a possibility that doesn't exists yet. It would also mean that you could lost this setting after an export/re-import. 

- Another thing that I would like to take into account, is the revamping of the data storage #16. The main idea is to make environments independents (each in its own data file), but what we are really heading toward is a system that looks more like a file editor (e.g. VSCode) where you can open an environment file from anywhere on your drive. What bothers me is that this system would maybe render the export feature a little bit obsolete, so, storing the network adapter, which is a "local" setting, in the environment file would de facto lead to conflicts when sharing through git or other systems. So we would need to store this in another data file `environment-xxxx.local.json`. I would prefer to wait for the new storage system before introducing a parallel system of local settings.

- You mentioned the CLI, but my understanding is that the CLI is mainly used on other machines, like CIs, so how would you ensure that the IP/adapter you chose when using the UI is also available on the CI/server machine? I wonder if this feature wouldn't be a good candidate for the CLI only, by adding a new flag like `--adapter x.x.x.x`.?

- this is the first time it has been requested, and as you said, it's probably not an essential feature. So, I would avoid adding new form elements on the main page, but rather move it to the environment settings. Also, the top bar will probably disappear in a near future (see https://github.com/mockoon/mockoon/issues/386#issuecomment-831071314 which is part of a bigger redesign effort)

I feel like this option raises a lot of questions if we want to do it well. But I am also basically thinking out loud and may be missing other ideas or solutions. Let me know what you think!

**Comment by barefootdeveloper** [comment]
Yeah, I think this is really a niche status feature as far as setting specific hostnames, more than anything I'd like the ability to set Mockoon servers to "Localhost Only" from the default of all adapters.   

When I was hacking and testing my changes I realized you could do interesting things like change the behavior based upon the hostname with proxies.

I'm okay waiting for a bigger vision. The last option I'd throw out there is something like this
![image](https://user-images.githubusercontent.com/3580448/118406063-c2fa0d80-b648-11eb-899e-e646c11eb561.png)

This would be shareable between users because 0.0.0.0 or 127.0.0.1 would work for everyone and then maybe we'd just need a flag for the cli to override the setting if we wanted to disable it.

**Comment by 255kb** [comment]
Would this be an alternative to the network adapter selector? It could be a quick win indeed.

**Comment by barefootdeveloper** [comment]
It would be yeah, it's what led me to wanting the change, I just went a little overboard :)

**Comment by 255kb** [comment]
Then I'm all for it :)
One detail, I am not sure about the location of the checkbox and keeping the "lock" as an icon for server settings but that's a detail and we can refine later.

**Comment by barefootdeveloper** [comment]
We think alike, I was actually thinking of https://fonts.google.com/icons?selected=Material+Icons+Outlined:storage just because it looks like a server. In font-awesome basically the same picture resolves to server.

https://fontawesome.com/icons/server?style=solid

Do you want me to modify the acceptance criteria of this issue or just create a new one to lose all the chatter?

**Comment by 255kb** [comment]
This icon looks fine indeed.
So I understand you want to work on this, or it's even already in progress? 🙂

**Comment by barefootdeveloper** [comment]
Yes please :)

**Comment by 255kb** [comment]
ok, I added it to the roadmap. 
Also no need to create another issue, at least we have all the history here.

**Comment by 255kb** [comment]
I merged everything, you will see that I made some small changes: 
- I migrated all the test environments.json files (using the script ./scripts/migrate-tests.js)
- I removed the two custom environments.json files you added for the tests and used the basic-data JSON file + UI manipulation to set `localhostonly` to `true` instead. I would like to avoid creating too many test files when it's not necessary, and instead manipulate the UI. It's also good to catch UI regressions. 
- When everything was running in the pipeline I found 3 tests not passing, 2 due to a missing return and one due to the `ip` helper not getting anymore the `::ffff:` prefix (this I have no clue why). 
Thanks again for the contribution!
